src/access/Subjects.h

```diff
@@ -0,0 +1,208 @@
+/*
+ *    Copyright (c) 2022 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <limits>
+
+#include <access/AuthMode.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/InPlace.h>
+#include <lib/core/NodeId.h>
+#include <lib/support/Variant.h>
+
+namespace chip {
+namespace Access {
+
+struct PaseSubject
+{
+    static constexpr uint16_t kPasscodeId = 0;
+    bool operator==(const PaseSubject & that) const { return true; }
+};
+
+struct NodeSubject
+{
+    NodeSubject(NodeId aNodeId) : nodeId(aNodeId) {}
+    NodeId nodeId;
+    bool operator==(const NodeSubject & that) const { return this->nodeId == that.nodeId; }
+};
+
+struct GroupSubject
+{
+    GroupSubject(GroupId aGroupId) : groupId(aGroupId) {}
+    GroupId groupId;
+    bool operator==(const GroupSubject & that) const { return this->groupId == that.groupId; }
+};
+
+class OperationalNodeId;
+
+/** A scoped subject is a unique identifier with-in a fabric scope. */
+class ScopedSubject
+{
+public:
+    ScopedSubject() {}
+
+    template <typename T, class... Args>
+    constexpr explicit ScopedSubject(InPlaceTemplateType<T>, Args &&... args) :
+        mSubject(InPlaceTemplate<T>, std::forward<Args>(args)...)
+    {}
+
+    template <typename T, typename... Args>
+    static ScopedSubject Create(Args &&... args)
+    {
+        return ScopedSubject(InPlaceTemplate<T>, std::forward<Args>(args)...);
+    }
+
+    Access::AuthMode GetAuthMode() const
```

Comment:
Raising the question of whether the Variant use here is worth it or whether a union and an `Access::AuthMode` discriminator would be simpler and smaller code....

Reply:
It is definitely better than `<Access::AuthMode, uint64_t (SubjectId)>`. for several reasons:

1. It can be compared against to en `OperationalNodeId` struct, which can be used by IM or other parts to identify a Node.
2. It is type-safe to retrieved `NodeId` or `GroupId` from it , without a cast.

Reply:
> It can be compared against to en OperationalNodeId struct,

That would be true anyway for this class, no?  The question is how it's implemented.

> It is type-safe to retrieved NodeId or GroupId from it , without a cast.

It would be safe anyway, with a union (which is not the same thing as `uint64_t`).  The only thing going on here is that the Variant is hiding the union, and we are building an adapter on top of it....

Reply:
A variant is a type-safe union, it is already better that `<type id, union>` tuple. The variant provides lots of runtime safety checks, which are also opt-out in production build. On the contract, union is more error prone.

Structs and classes in union are restricted, especially can not have constructors, and const fields, which requires more line of codes (and maybe more bugs).

I have also added `WildcardSubject` and `CaseAuthenticatedTagSubject`, whose implementation using union is as straightforward as using variant. With variant, just add the corresponding struct, and add the new struct type into variant definition.